### PR TITLE
[norman] Round-2: SE(3) local-frame coord features (A02)

### PR DESCRIPTION
**Pull request by morganmcg1**
## Hypothesis

Augment surface and volume point features with **SE(3)-invariant local-frame
coordinates**: for each point, compute the offset between the point and the
centroid of its k=16 nearest neighbours, then decompose that offset into
normal and tangential components using the point's surface normal. This
replaces absolute Cartesian coordinates with geometry-relative ones that are
invariant to rigid-body translations and partially equivariant to rotations.

**Why this should work:**
- The Transolver currently receives raw `[x, y, z, nx, ny, nz, area]` surface
  features — absolute positions in a global coordinate frame. A 5 mm shift of
  the entire car geometry should not change aerodynamic predictions, but the
  model must learn this from scratch across only 400 training geometries.
- Wall shear stress (`tau_x/y/z`) transforms as a pseudo-vector under rotation,
  so expressing point positions in a local surface-tangent frame directly reduces
  the model's rotation-learning burden for the hardest axes.
- **Prior art:** noam branch PR #2377 (MERGED) confirmed a robust gain from
  SE(2)-equivariant local frame encoding on TandemFoil. DrivAerML has full 3D
  orientation variance (yaw, road clearance), so the 3D generalisation should
  be at least as effective.
- This is a feature preprocessing change only — the Transolver backbone is
  completely unchanged.

**Primary targets:** `wall_shear_x/y/z` (largest gap to AB-UPT:
tau_y 19.49 vs target 3.65; tau_z 21.12 vs target 3.63).

### Step 1: Implement the `LocalFrameEncoder`

Add this class to `train.py` (before the main model):

```python
import torch
import torch.nn.functional as F

class LocalFrameEncoder(torch.nn.Module):
    """
    Augments point features with local-frame relative coordinates.
    For each point, finds its k nearest neighbours, computes the offset
    to the local centroid, and decomposes it into:
      - normal component (along surface normal n)
      - tangential component (magnitude of tangential part)
      - total distance to centroid
    These three scalars are appended to the original feature vector.
    Input:  [B, N, F]  with positions in features[:, :, :3] and normals in features[:, :, 3:6]
    Output: [B, N, F+3]
    """
    def __init__(self, k: int = 16):
        super().__init__()
        self.k = k

    def forward(self, x: torch.Tensor) -> torch.Tensor:
        # x: [B, N, F]; first 3 dims are xyz, dims 3:6 are normals
        pos = x[:, :, :3]       # [B, N, 3]
        n   = x[:, :, 3:6]      # [B, N, 3]  (surface normals, already normalised)
        n   = F.normalize(n, dim=-1)

        # k-NN in position space (batched pairwise distance)
        # For 65k pts this is O(N^2) — chunk if needed, but typically fast with fp16
        with torch.no_grad():
            dist2 = torch.cdist(pos, pos)                    # [B, N, N]
            _, idx = dist2.topk(self.k + 1, largest=False)   # [B, N, k+1]
            idx = idx[:, :, 1:]                              # drop self: [B, N, k]

        # Gather neighbour positions and compute centroid
        B, N, _ = pos.shape
        idx_exp = idx.unsqueeze(-1).expand(B, N, self.k, 3)  # [B, N, k, 3]
        nbr_pos = torch.gather(pos.unsqueeze(2).expand(B, N, N, 3), 2, idx_exp)
        centroid = nbr_pos.mean(dim=2)                        # [B, N, 3]

        # Decompose offset in local surface frame
        offset = pos - centroid                               # [B, N, 3]
        n_comp  = (offset * n).sum(dim=-1, keepdim=True)     # [B, N, 1]  normal component
        tan_vec = offset - n_comp * n                         # [B, N, 3]  tangential part
        tan_mag = tan_vec.norm(dim=-1, keepdim=True)          # [B, N, 1]  tangential magnitude
        dist    = offset.norm(dim=-1, keepdim=True)           # [B, N, 1]  total distance

        return torch.cat([x, n_comp, tan_mag, dist], dim=-1)  # [B, N, F+3]
```

**Memory note:** the `torch.cdist` call at 65k points will require ~30 GB at fp32.
Wrap it in `with torch.cuda.amp.autocast(enabled=False):` and consider computing
it in chunks of 8–16k points if you see OOM:

```python
# Chunked k-NN (drop-in for the cdist block above)
chunk = 8192
idx_list = []
for i in range(0, N, chunk):
    d = torch.cdist(pos[:, i:i+chunk], pos)  # [B, chunk, N]
    _, k_idx = d.topk(self.k+1, largest=False)
    idx_list.append(k_idx[:, :, 1:])
idx = torch.cat(idx_list, dim=1)  # [B, N, k]
```

### Step 2: Wire it into the model

In the model's `__init__`, instantiate the encoder and update input dimensions:

```python
self.local_frame = LocalFrameEncoder(k=16)

# Surface head input changes: 7 → 10 features
# Find the first Linear layer in the surface encoding path and change:
# nn.Linear(7, hidden_dim)  →  nn.Linear(10, hidden_dim)

# Volume head input changes: 4 → 7 features  
# nn.Linear(4, hidden_dim)  →  nn.Linear(7, hidden_dim)
```

**For volume points:** the local-frame encoder needs normals, which volume points
don't have. Use a simpler version: just append the offset to the k-NN centroid
in global frame (3D vector), without decomposing into normal/tangential components:

```python
# Volume version (no normals available):
vol_offset = vol_pos - vol_nbr_centroid          # [B, N_vol, 3]
vol_x_aug = torch.cat([vol_x, vol_offset], dim=-1)  # [B, N_vol, 4+3] = [B, N_vol, 7]
```

### Step 3: Apply in `forward`

```python
# Before encoding surface points:
surface_x = self.local_frame(surface_x)   # [B, N_surf, 7] → [B, N_surf, 10]

# Before encoding volume points (simplified version):
vol_x = self.vol_local_offset(vol_x)      # [B, N_vol, 4] → [B, N_vol, 7]
```

### Step 4: Run the experiment

```bash
cd target/
python train.py \
  --volume-loss-weight 2.0 \
  --batch-size 8 \
  --validation-every 1 \
  --use-tangential-wallshear-loss \
  --lr 2e-4 --weight-decay 5e-4 \
  --train-surface-points 65536 --eval-surface-points 65536 \
  --train-volume-points 65536 --eval-volume-points 65536 \
  --model-layers 4 --model-hidden-dim 256 --model-heads 4 --model-slices 128 \
  --ema-decay 0.9995 \
  --ema-decay-start 0.99 --ema-decay-end 0.9999 \
  --gradient-log-every 100 --weight-log-every 100 --no-log-gradient-histograms \
  --wandb-group norman-se3-local-frame
```

Note: `--ema-decay-start 0.99 --ema-decay-end 0.9999` uses the cosine EMA schedule
from your PR #13 — this will be on `yi` once that PR's rebase completes.
`--use-tangential-wallshear-loss` is already on `yi` (PR #11, default off).

**If k-NN computation at 65k points is too slow:** try `--train-surface-points 40000
--eval-surface-points 40000` to reduce N. The local-frame augmentation is the
hypothesis, not the point count.

## Baseline

Current verified yi best: **PR #13 (your own cosine EMA run)**, `wio9pqw2`,
4 epochs, best_epoch=4. Beat 15.82 on `abupt_axis_mean_rel_l2_pct` to set a new yi best.

| Metric | yi verified best (PR #13 pending) | yi merged (PR #9) | AB-UPT target |
|---|---:|---:|---:|
| `abupt_axis_mean_rel_l2_pct` | **15.82** | 17.39 | — |
| `surface_pressure_rel_l2_pct` | **9.99** | 11.07 | 3.82 |
| `wall_shear_rel_l2_pct` | **16.60** | 18.32 | 7.29 |
| `volume_pressure_rel_l2_pct` | **14.21** | 15.21 | 6.08 |
| `wall_shear_x_rel_l2_pct` | **14.27** | 15.65 | 5.35 |
| `wall_shear_y_rel_l2_pct` | **19.49** | 21.86 | 3.65 |
| `wall_shear_z_rel_l2_pct` | **21.12** | 23.18 | 3.63 |

**Reproduce (PR #13 + PR #9 base config):**
```bash
cd target/
python train.py \
  --volume-loss-weight 2.0 --batch-size 8 --validation-every 1 \
  --lr 2e-4 --weight-decay 5e-4 \
  --train-surface-points 65536 --eval-surface-points 65536 \
  --train-volume-points 65536 --eval-volume-points 65536 \
  --model-layers 4 --model-hidden-dim 256 --model-heads 4 --model-slices 128 \
  --ema-decay 0.9995 --ema-decay-start 0.99 --ema-decay-end 0.9999 \
  --gradient-log-every 100 --weight-log-every 100 --no-log-gradient-histograms
```

## What to report

- W&B run ID and full `test_primary/*` table.
- Peak GPU memory (to verify k-NN did not OOM).
- Per-epoch val trajectory — particularly whether `tau_y` and `tau_z` improve relative to the baseline (these are the primary targets for the local-frame hypothesis).
- Whether the chunked k-NN was needed.
- Parameter count delta (should be near zero — only the first Linear input_dim changes).

### CODE REVIEW

**Comment by morganmcg1** [comment]
OPERATOR: Pausing this follow-up assignment temporarily so student:norman can first resolve the active dependency PR #13 rebase/merge blockage. Removing student:norman from this PR; the advisor should reassign this local-frame follow-up after #13 lands on yi.

**Comment by morganmcg1** [comment]
OPERATOR UPDATE: Using the cleaner paused state for this follow-up: keeping student:norman for ownership, but removing status:wip so it is not picked up until PR #13 lands.

**Comment by morganmcg1** [comment]
ADVISOR: Closing PR #28 because Closing this draft — it was designed against the PR #13 baseline (15.82). The current yi best is 12.74 (PR #66, 6L/256d with per-axis tau upweighting). The SE(3) local-frame hypothesis remains interesting for the wall_shear_y/z gap (still 4.1-4.2× AB-UPT), but the experiment needs to be redesigned for the current baseline config (6L/256d, lr=2e-4, bs=8, clip=1.0, wallshear-y-weight=2.0, wallshear-z-weight=2.0). The k-NN at 65k points memory concern noted here still applies. If norman wants to revisit, open a new Round-3 PR with the updated config..
